crates/proto/src/quic/quic_client_stream.rs

```diff
@@ -0,0 +1,295 @@
+        self.crypto_config = crypto_config;
+        self
+    }
+
+    /// Sets the address to connect from.
+    pub fn bind_addr(&mut self, bind_addr: SocketAddr) -> &mut Self {
+        self.bind_addr = Some(bind_addr);
+        self
+    }
+
+    /// Sets a good set of defaults for the DoQ transport config
+    pub fn default_transport_config(&mut self) -> &mut Self {
+        self.transport_config(TransportConfig::default())
+    }
+
+    /// This will override the max_concurrent_bidi_streams and max_concurrent_uni_streams to 0, as DoQ doesn't support server push
+    pub fn transport_config(&mut self, mut transport_config: TransportConfig) -> &mut Self {
+        sanitize_transport_config(&mut transport_config);
+        self.transport_config = Arc::new(transport_config);
+
+        self
+    }
+
+    /// Creates a new QuicStream to the specified name_server
+    ///
+    /// # Arguments
+    ///
+    /// * `name_server` - IP and Port for the remote DNS resolver
+    /// * `dns_name` - The DNS name, Subject Public Key Info (SPKI) name, as associated to a certificate
+    pub fn build(self, name_server: SocketAddr, dns_name: String) -> QuicClientConnect {
+        QuicClientConnect(Box::pin(self.connect(name_server, dns_name)) as _)
+    }
+
+    async fn connect(
+        self,
+        name_server: SocketAddr,
+        dns_name: String,
+    ) -> Result<QuicClientStream, ProtoError> {
+        let connect = if let Some(bind_addr) = self.bind_addr {
+            <tokio::net::UdpSocket as UdpSocket>::connect_with_bind(name_server, bind_addr)
+        } else {
+            <tokio::net::UdpSocket as UdpSocket>::connect(name_server)
+        };
+
+        let socket = connect.await?;
+        let socket = socket.into_std()?;
+
+        let (mut endpoint, _incoming) = Endpoint::new(EndpointConfig::default(), None, socket)?;
+
+        // ensure the ALPN protocol is set correctly
+        let mut crypto_config = self.crypto_config;
+        crypto_config.alpn_protocols = vec![quic_stream::DOQ_ALPN.to_vec()];
+
+        let client_config = ClientConfig::new(Arc::new(crypto_config));
+        endpoint.set_default_client_config(client_config);
+
+        let connecting = endpoint.connect(name_server, &dns_name)?;
+        // TODO: for Client/Dynamic update, don't use RTT, for queries, do use it.
+
+        let connection = connecting.await?;
+        let NewConnection {
+            connection: quic_connection,
+            ..
+        } = connection;
+
+        Ok(QuicClientStream {
+            quic_connection,
+            name_server_name: Arc::from(dns_name),
+            name_server,
+            is_shutdown: false,
+        })
+    }
+}
+
+fn sanitize_transport_config(transport_config: &mut TransportConfig) {
+    transport_config.max_concurrent_bidi_streams(VarInt::from_u32(0));
+    transport_config.max_concurrent_uni_streams(VarInt::from_u32(0));
+}
+
+fn client_config_tls12_webpki_roots() -> TlsClientConfig {
```

Comment:
This function name is surprising, since QUIC doesn't support TLS 1.3. Maybe you should extract the change to allow other TLS versions in the other protocols into a separate PR and also change the function names (assuming this is a copy)?

Reply:
oh, this might be unnecessary, I think defaults are probably fine... I can definitely move the other stuff into a separate PR.